python/ray/data/dataset.py

```diff
@@ -4055,52 +4062,53 @@ def __iter__(self):
         pipe = DatasetPipeline(it, False, length=len(it._splits))
         if read_stage:
             pipe = pipe.foreach_window(
-                lambda ds, read_stage=read_stage: Dataset(
+                lambda ds, read_stage=read_stage: Datastream(
                     ds._plan.with_stage(read_stage), ds._epoch, True
                 )
             )
         return pipe
 
-    @Deprecated(message="Use `Dataset.cache()` instead.")
-    def fully_executed(self) -> "Dataset[T]":
+    @Deprecated(message="Use `Datastream.cache()` instead.")
+    def fully_executed(self) -> "MaterializedDatastream[T]":
         logger.warning(
             "The 'fully_executed' call has been renamed to 'cache'.",
         )
         return self.cache()
 
-    @Deprecated(message="Use `Dataset.is_cached()` instead.")
+    @Deprecated(message="Use `Datastream.is_cached()` instead.")
     def is_fully_executed(self) -> bool:
         logger.warning(
             "The 'is_fully_executed' call has been renamed to 'is_cached'.",
         )
         return self.is_cached()
 
     def is_cached(self) -> bool:
-        """Returns whether this Dataset has been cached in memory.
+        """Returns whether this datastream has been cached in memory.
 
         This will return False if the output of its final stage hasn't been computed
         yet.
         """
         return self._plan.has_computed_output()
 
     @ConsumptionAPI(pattern="store memory.", insert_after=True)
-    def cache(self) -> "Dataset[T]":
-        """Evaluate and cache the blocks of this Dataset in object store memory.
+    def cache(self) -> "MaterializedDatastream[T]":
+        """Evaluate and cache the blocks of this datastream in object store memory.
 
-        This can be used to read all blocks into memory. By default, Datasets
+        This can be used to read all blocks into memory. By default, Datastreams
         doesn't read blocks from the datasource until the first transform.
 
         Returns:
-            A Dataset with all blocks fully materialized in memory.
+            A Datastream with all blocks fully materialized in memory.
         """
-        self._plan.execute(force_read=True)
-        return self
+        copy = Datastream.copy(self, _as=MaterializedDatastream)
+        copy._plan.execute(force_read=True)
+        return copy
```

Comment:
I feel returning a new object would make the semantic a little confusing. `ds.cache()` is having side effect on `ds`, but `ds` itself is still a `Datastream` with `is_cached() == True`. Feeling overloaded that we have two objects

* `Datastream` with `is_cached() == True` (this is quite weird)
* `MaterializedDatastream`

```
>>> ds = ray.data.range(10)
>>> output_ds = ds.cache()
>>> output_ds
MaterializedDatastream(num_blocks=10, num_rows=10, schema=<class 'int'>)
>>> ds.is_cached()
True
>>> ds
Datastream(num_blocks=10, num_rows=10, schema=<class 'int'>)
```

Reply:
I'll try to fix it so there isn't a side effect. That was not intended originally.

Reply:
Fixed by using a deep-copy.